### PR TITLE
DateStep time for moscow fixes #7261

### LINKED ISSUE #7261: Date element does not validate in some timezones
I use a date element in my form. It doesn't validate because of DateStep validator. DateStep validator is a default validator in date element in ZF2. When I submit a form I have this validation error: "The input is not a valid step"
Here is a configuration of my Date element in form init() function:
$this->add([
    'name' => 'test',
    'type' => 'Date',
    'options' => [
        'label' => 'test',
    ],
]);
As you can see there is nothing special in this configuration.
Here is an example with incorrect work of DateStep validator:
date_default_timezone_set('Europe/Moscow');
$validator = new \Zend\Validator\DateStep(array(
    'format'    => 'd-m-Y',
    'baseValue' => date('d-m-Y', 0),
    'step'      => new \DateInterval("P1D"),
));
$result = $validator->isValid('12-08-2014'); // $result == false
This validation error is because DateStep validator use DateTime::diff() method. Here is an another example:
date_default_timezone_set('Europe/Moscow');
$datetime1 = new DateTime('2014-08-12');
$datetime2 = new DateTime('1970-01-01');
$interval = $datetime1->diff($datetime2);
In this example $interval has 44 years, 7 months, 10 days, 23 hours(!), 0 minutes, 0. 23 hours is the reason why DateInterval validator works incorrectly

### PR DESCRIPTION
**Pull request by Victory**
Here is a test and fix for the "Wintertime" bug for DateStepValidator in Moscow.

### CODE REVIEW

**Comment by Victory** [comment]
As far as I can tell, the failing build now has nothing to do with my commits.

File: library/Zend/Validator/DateStep.php [hunk 1]
```diff
@@ -376,6 +381,26 @@ private function fallbackIncrementalIterationLogic(
...
     /**
+     * Returns true if the interval spans a special case time,
+     * such as days in Europe/Moscow time 26-03-2011
+     *
+     * @param DateTime     $baseDate
+     * @param DateTime     $valueDate
+     *
+     * @return bool
+     */
+    private function isSpaningSpecialCase($baseDate, $valueDate)
```
Comment (weierophinney) [inline library/Zend/Validator/DateStep.php]: `s/Spaning/Spanning/`

File: library/Zend/Validator/DateStep.php [hunk 2]
```diff
@@ -219,6 +219,11 @@ public function isValid($value)
         $timeDiff  = $valueDate->diff($baseDate, true);
         $diffParts = array_combine($unitKeys, explode('|', $timeDiff->format('%y|%m|%d|%h|%i|%s')));
 
+        // check if this is a special case where time is added or lost
+        if ($this->isSpaningSpecialCase($baseDate, $valueDate)) {
```
Comment (weierophinney) [inline library/Zend/Validator/DateStep.php]: `s/Spaning/Spanning/`

File: library/Zend/Validator/DateStep.php [hunk 3]
```diff
@@ -376,6 +381,26 @@ private function fallbackIncrementalIterationLogic(
...
+     *
+     * @param DateTime     $baseDate
+     * @param DateTime     $valueDate
+     *
+     * @return bool
+     */
+    private function isSpaningSpecialCase($baseDate, $valueDate)
+    {
+        // Check for missing hour in Europe/Moscow time
+        if ($this->getTimezone() == new \DateTimeZone("Europe/Moscow")) {
```
Comment (weierophinney) [inline library/Zend/Validator/DateStep.php]: Import `DateTimeZone`, please.

File: tests/ZendTest/Validator/DateStepTest.php [hunk 4]
```diff
@@ -120,4 +123,30 @@ public function testStepError()
...
+    /**
+     * @dataProvider moscowWinterTimeDataProvider
+     */
+    public function testMoscowWinterTime($dateToValidate)
+    {
+        $validator = new \Zend\Validator\DateStep(array(
+            'format' => 'd-m-Y',
+            'baseValue' => date('d-m-Y', 0),
+            'step' => new DateInterval("P1D"),
+            'timezone' => new \DateTimeZone('Europe/Moscow')
```
Comment (weierophinney) [inline tests/ZendTest/Validator/DateStepTest.php]: Import `DateTimeZone`, please.

**Comment by weierophinney** [comment]
@Victory please address the feedback left by both myself and @froschdesign , and I'll review again for merge.

**Comment by Victory** [comment]
@weierophinney - I have fixed the typo and used `use`.

The discussion with @froschdesign is on going.

Thanks!

**Comment by Victory** [comment]
@weierophinney and @froschdesign  i have just change my approach quite a bit to not have any checks for specific time zones. I think this is much better approach then what i tired before.

**Comment by froschdesign** [comment]
@Victory 
Much better!

**Comment by Victory** [comment]
Ok @weierophinney I think I have addressed all concerns for this pull request. Let me know if there is anything else you would like updated.

**Comment by weierophinney** [comment]
Merged to develop for release with 2.4.

**Pull request closed without merging**